strawberry/permission.py

```diff
@@ -89,6 +111,85 @@ class AutoDirective:
 
         return self._schema_directive
 
+    @cached_property
+    def is_async(self) -> bool:
+        return iscoroutinefunction(self.has_permission)
+
+    def __and__(self, other: BasePermission):
+        return AndPermission([self, other])
+
+    def __or__(self, other: BasePermission):
+        return OrPermission([self, other])
+
+
+class CompositePermission(BasePermission, abc.ABC):
+    def __init__(self, child_permissions: List[BasePermission]):
+        self.child_permissions = child_permissions
+
+    def on_unauthorized(self, **kwargs: dict) -> Any:
+        failed_permissions = kwargs.get("failed_permissions", [])
+        for permission in failed_permissions:
+            permission.on_unauthorized()
+
+    @cached_property
+    def is_async(self) -> bool:
+        return any(x.is_async for x in self.child_permissions)
+
+
+class AndPermission(CompositePermission):
+    def has_permission(
+        self, source: Any, info: Info, **kwargs: dict
+    ) -> Union[bool, Awaitable[bool], (False, dict), Awaitable[(False, dict)]]:
+        if self.is_async:
+            return self._has_permission_async(source, info, **kwargs)
+
+        for permission in self.child_permissions:
+            if not permission.has_permission(source, info, **kwargs):
+                return False, {"failed_permissions": [permission]}
+        return True
+
+    async def _has_permission_async(
+        self, source: Any, info: Info, **kwargs: dict
+    ) -> Union[bool, (False, dict)]:
+        for permission in self.child_permissions:
+            if not await await_maybe(permission.has_permission(source, info, **kwargs)):
+                return False, {"failed_permissions": [permission]}
+        return True
+
+    def __and__(self, other: BasePermission):
+        self.child_permissions.append(other)
```

Comment:
I think `__and__` and `__or__` should return a new instance, because it shouldn't mutate the object itself

Reply:
I feel like that would create a lot of heap waste if you have `A or B or C or D` in a lot of fields. Is it actually in the python spec that and / or shouldn't mutate the object itself? I only found this https://docs.python.org/3/reference/datamodel.html#emulating-numeric-types

Reply:
Since there's `|=` operator I think it shouldn't 🤔 numeric types are kind of special because they're never mutated

Reply:
Agreed it seems weird to me to mutate objects type.  It's a side-effect that you wouldn't really expect to have to plan around

Reply:
You are 100 % right, i only thought of ad hoc use cases when using the operators on a field extension, not about pre-composed permissions